### PR TITLE
Use object_type methods to correctly construct strawberry fields

### LINKED ISSUE #1109: Type resolver not called when extending from pydantic model
I have 2 types, Category and Articles, and I want the Category type to return its related Articles(I have tried with other simpler types too), but when defining a resolver in the Category type its not being called:

```python
def articles_helper(root):  
    return []      #  <-whatever I return here, it always shows null
   
   
@strawberry.experimental.pydantic.type(
    model=Category,
    fields=[
        'title',
        'description',
        'content',
        'template',
        'slug',
        'date_published',
        'date_updated',
    ])
class CategoryType(Category):
    id: int
    articles: Optional[List[ArticleType]] = strawberry.field(resolver=articles_helper)  # <- resolver not called
```    

Query type:
```python
@strawberry.type()
class Query:
    category: Optional[CategoryType] = field(resolver=get_category)  # this resolver is OK
```

Always returns `null`:
```
Query:

{
  category(categoryId: 1) {
    id
    title
    articles {
      title
    }
  }
}

Response:

{
  "data": {
    "category": {
      "id": 1,
      "title": "test",
      "articles": null      <- always
    }
  }
}
```

**I works when I don't extend from a pydantic model**

### PR DESCRIPTION
**Pull request by jkimbo**
## Description

Fix for #1109

Note: This PR also changes how we avoid the issue where a field without a default value is defined after a field with a default value. The fix I've implemented is to sort the fields. This will change the order in which they are shown in the schema but I don't think it's a big issue. 

## Types of Changes

- [ ] Core
- [x] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

* Fixes #1109

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by patrick91** [comment]
> This will change the order in which they are shown in the schema but I don't think it's a big issue.

Preserving the schema order is useful sometimes, ie apollo federation uses that to check types are the same (not sure it makes a lot of sense to be honest though)

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/), so make sure to pick the appropriate type. If in doubt feel free to ask :)

**Comment by jkimbo** [comment]
> > This will change the order in which they are shown in the schema but I don't think it's a big issue.
> 
> Preserving the schema order is useful sometimes, ie apollo federation uses that to check types are the same (not sure it makes a lot of sense to be honest though)

Yeah it's not ideal but to be honest the best way of resolving it is to replace dataclasses with something that doesn't require fields without defaults to be defined before fields with.

File: strawberry/experimental/pydantic/object_type.py [hunk 1]
```diff
@@ -91,34 +92,36 @@ def wrap(cls):
...
+                    field,
                 )
-                for name, type_ in cls_annotations.items()
+                for field in extra_fields
             )
         )
 
+        # Sort fields so that fields with missing defaults go first
+        # because dataclasses require that fields with no defaults are defined
+        # first
```
Comment (patrick91) [inline strawberry/experimental/pydantic/object_type.py:111]: should we add a test for this case?

Comment (jkimbo) [reply strawberry/experimental/pydantic/object_type.py:111]: The changes I made to `test_basic` already cover this I think (the order changed)

Comment (patrick91) [reply strawberry/experimental/pydantic/object_type.py:111]: oh man, I just realised this is just for Pydantic :D

**Pull request merged**